nuclear/message/Neutron.cmake

```diff
@@ -1,274 +1,193 @@
-# Get the relative path to our message directory
-get_filename_component(message_include_dir "${PROJECT_SOURCE_DIR}/${NUCLEAR_MESSAGE_DIR}/.." ABSOLUTE)
-file(RELATIVE_PATH message_include_dir ${PROJECT_SOURCE_DIR} ${message_include_dir})
+# Get the path to our parent directory above the message folder
+get_filename_component(message_parent_dir "${PROJECT_SOURCE_DIR}/${NUCLEAR_MESSAGE_DIR}/.." ABSOLUTE)
 
-# Get our two include directories for message
-set(message_source_include_dir "${PROJECT_SOURCE_DIR}/${message_include_dir}")
-set(message_binary_include_dir "${PROJECT_BINARY_DIR}/${message_include_dir}")
+# Generate the protocol buffers that are built into NUClear
+set(builtin_dir "${CMAKE_CURRENT_SOURCE_DIR}/proto")
+set(message_dir "${PROJECT_SOURCE_DIR}/${NUCLEAR_MESSAGE_DIR}")
+file(GLOB_RECURSE builtin_protobufs "${builtin_dir}/**.proto")
+file(GLOB_RECURSE message_protobufs "${message_dir}/**.proto")
 
-# Make our message include directories variable
-set(NUCLEAR_MESSAGE_INCLUDE_DIRS
-    ${CMAKE_CURRENT_SOURCE_DIR}/include ${message_source_include_dir} ${message_binary_include_dir}
-    CACHE INTERNAL "Include directories for the message folder and generated sources"
-)
+# Locations to store each of the output components
+set(pb_out "${CMAKE_CURRENT_BINARY_DIR}/protobuf")
+set(nt_out "${CMAKE_CURRENT_BINARY_DIR}/neutron")
+set(py_out "${CMAKE_CURRENT_BINARY_DIR}/python")
 
-# Include our message directories
-include_directories(${NUCLEAR_MESSAGE_INCLUDE_DIRS})
-
-# Get our source and binary directories for message
-set(message_source_dir "${PROJECT_SOURCE_DIR}/${NUCLEAR_MESSAGE_DIR}")
-set(message_binary_dir "${PROJECT_BINARY_DIR}/${NUCLEAR_MESSAGE_DIR}")
+# Files that are used to generate the neutron files
+file(GLOB_RECURSE message_class_generator_files "${CMAKE_CURRENT_SOURCE_DIR}/generator/**.py")
 
 # We need protobuf and python to generate the neutron messages
 find_package(Protobuf REQUIRED)
 find_package(PythonInterp 3 REQUIRED)
 
-include_directories(SYSTEM ${Protobuf_INCLUDE_DIRS})
-
-# If we have the package pybind11 we can use to go generate python bindings FIND_PACKAGE(pybind11)
-
-# We need Eigen3
+# We need Eigen for neutron messages
 find_package(Eigen3 REQUIRED)
-include_directories(SYSTEM ${Eigen3_INCLUDE_DIRS})
 
-# We need Eigen3
-find_package(Eigen3 REQUIRED)
-include_directories(SYSTEM ${Eigen3_INCLUDE_DIRS})
-
-# If we found pybind11 include its directories
+# If we have pybind11 we need to generate our final binding class
 if(pybind11_FOUND)
   find_package(PythonLibsNew 3 REQUIRED)
 
   include_directories(SYSTEM ${pybind11_INCLUDE_DIRS})
   include_directories(SYSTEM ${PYTHON_INCLUDE_DIRS})
 endif()
```

Comment:
Also all that include directories stuff is totally not modern cmake

Reply:
So if I add a `find_package(pybind11)`, then it complains that toolchainlibraryfinder tries to include directories for something we don't build in this line (86):
 `target_include_directories(${PACKAGE_NAME}::${PACKAGE_NAME} SYSTEM INTERFACE ${${PACKAGE_NAME}_INCLUDE_DIR})`, 
but it's not a target of ours obviously.
To fix that, I tried adding ` LIBRARY pybind11` to where `toolchainlibraryfinder` is called in `Findpybind11.cmake`:
```
include(ToolchainLibraryFinder)
ToolchainLibraryFinder(NAME pybind11 HEADER pybind11/pybind11.h LIBRARY pybind11)
#EOF
```
This kind of works, but not really. When it tries to find `pybind11`, it fails:
`-- Could NOT find pybind11 (missing: pybind11_LIBRARY pybind11_INCLUDE_DIR)`
but yet it still sets `pybind11_FOUND`, and it proceeds to complain inside that `if(pybind11_FOUND)` that `  target_include_directories(nuclear_message SYSTEM ${pybind11_INCLUDE_DIRS})` isn't valid, presumably because `pybind11_INCLUDE_DIRS` isn't set, because pybind11 wasn't actually found:
```cmake
-- Found PythonInterp: /usr/sbin/python3 (found suitable version "3.9.1", minimum required is "3") 
-- Could NOT find pybind11 (missing: pybind11_LIBRARY pybind11_INCLUDE_DIR) 
-- Found PythonLibs: /usr/lib64/libpython3.9.so
CMake Error at nuclear/message/Neutron.cmake:183 (target_include_directories):
  target_include_directories called with invalid arguments
Call Stack (most recent call first):
  nuclear/message/CMakeLists.txt:6 (include)
```

Not exactly sure how I should proceed from here. 
Also, could you please elaborate on what you mean by "will need to be wrapped in modernising things too" @Bidski?